### PR TITLE
Specify that server names cannot be %-encoded

### PR DESCRIPTION
**Pull request by richvdh**
This is spelling out the current situation.

Fixes https://github.com/matrix-org/matrix-doc/issues/1368.

### CODE REVIEW

File: specification/appendices/identifier_grammar.rst [hunk 1]
```diff
@@ -23,13 +23,37 @@ A homeserver is uniquely identified by its server name. This value is used in a
...
+    IPv6char    = DIGIT / %x41-46 / %x61-66 / ":" / "."
+                      ; 0-9, A-F, a-f, :, .
+
+    dns-name    = *255dns-char
+
+    dns-char    = DIGIT / ALPHA / "-" / "."
+
+— in other words, the server name is the hostname, followed by an optional
+numeric port specifier. The hostname may be a dotted-quad IPv4 address literal,
+an IPv6 address literal surrounded with square brackets, or a DNS name.
```
Comment (turt2live) [inline specification/appendices/identifier_grammar.rst]: We dropped "DNS name" across the spec in favour of "hostname". Might actually make more sense here though...

File: specification/appendices/identifier_grammar.rst [hunk 2]
```diff
@@ -23,13 +23,37 @@ A homeserver is uniquely identified by its server name. This value is used in a
...
+
+IPv4 literals must be a sequence of four decimal numbers in the
+range 0 to 255, separated by ``.``. IPv6 literals must be as specified by
+`RFC3513, section 2.2 <https://tools.ietf.org/html/rfc3513#section-2.2>`_.
+
+DNS names for use with Matrix should follow the conventional restrictions for
```
Comment (turt2live) [inline specification/appendices/identifier_grammar.rst]: another mention of DNS name (see previous comment)

File: specification/appendices/identifier_grammar.rst (hunk 1 above)
Comment (richvdh) [reply specification/appendices/identifier_grammar.rst]: my impression is that we have used "hostname" to mean "DNS name or IP literal", as this doc does.

Comment (richvdh) [reply specification/appendices/identifier_grammar.rst]: In fact I am going to update the BNF to use `hostname` here too.

File: specification/server_server_api.rst [hunk 3]
```diff
@@ -71,39 +71,40 @@ This version of the specification is generated from
...
+
+  Homeservers may use SRV records to load balance requests between multiple TLS
+  endpoints or to failover to another endpoint if an endpoint fails.
+
+When making requests to servers, use the hostname of the target server in the
+``Host`` header, regardless of the any hostname given in the SRV record. For
+example, if the server name is ``example.org``, and the SRV record resolves to
+``matrix.example.org``, the ``Host`` header in the request should be
+``example.org``. The port number for target server should **not** appear in the
+``Host`` header.
```
[the remark this answers is not in the document]
Comment (richvdh) [reply specification/server_server_api.rst]: apparently not. PTAL if you want to check the correction.

**Pull request merged**